datajoint/relational_operand.py

```diff
@@ -467,7 +468,12 @@ def __len__(self):
         """
         number of tuples in the relation.
         """
-        return U().aggr(self, n='count(*)').fetch1('n')
+        return self.connection.query(
+            'SELECT ' + (
+                'count(DISTINCT `{pk}`)'.format(pk='`,`'.join(self.primary_key)) if self.distinct else 'count(*)') +
```

Comment:
While this is a definite improvement, I seem to recall`count(*)` can be much slower than `count(some-indexed-or-just-small-field)` in some cases since `count` is an aggregation function performed on the lower-level result set.

Unfortunately I'm having trouble locating any good queries to verify this with (need to have dj generate sufficient joins where the `count(*)` operates on a subquery I believe). I _think_ the lower level implication being that a subselect requires creating a temporary table to hold the `*` data, on which an index lookup is done to compute the result; in cases where `*` is sufficiently small, this temporary table is stored in-memory (`< tmp_table_size`), in other cases, it is written to disk. If only some small field is used, the resulting temporary table is smaller and therefore fits in memory. Even for those cases where an on-disk table is used, a smaller dataset would require less transient data copy. I don't think there is anything to be done for the case of row-scans on large tables (e.g when using a `WHERE` clause). I haven't pinned down all of these things 100% however.

In any event, could the same pkey-splicing logic be used more generally (theoretically just using the first pkey field would work)?

This could potentially be a topic for further investigation; the current change does imply some improvement and I'm not concrete enough about the details here to entirely say for sure.

Reply:
Very good point brought up. It does appear that the `pkey` subselection could already be a general thing and I have been using this implementation of count locally to get a speed ups for all counts operation.

Reply:
With a subquery, you are correct, perhaps count(*) causes the full subquery result to be generated rather than just inferring how many rows it should have.  This would require a principally new approach, which we are working on now.

Reply:
Indeed - I don't care to have this addressed in this PR. Let's just keep a note of this and other queries with known performance issues for future improvements.

Reply:
Yes, the upcoming redesign is based on a new approach, leading to fewer subqueries.  But it's a major redesign and I would like to roll it into rev 0.11.0